# EXAMEN NLP M2 SIR

### Papa Sidy Mactar TRAORE M1 SI

## I- Questions de cours

### a. Qu'est-ce que le NLP (traitement automatique du langage naturel) et en quoi est-il important dans le domaine de l'informatique ?

Le traitement automatique du langage naturel (NLP) est un domaine de l'informatique, de l'intelligence artificielle et de la linguistique computationnelle qui s'intéresse aux interactions entre les ordinateurs et les langues humaines (naturelles). Il vise à permettre aux ordinateurs de comprendre, d'interpréter et de générer du texte en langage naturel d'une manière qui soit significative. Le traitement automatique du langage naturel (NLP) est crucial dans le domaine de l'informatique pour plusieurs raisons :

- Compréhension profonde du langage : Les systèmes NLP sont conçus pour identifier et comprendre la structure et le sens des mots, des phrases, des textes et des conversations. Ils doivent avoir une compréhension profonde et large du langage pour être utiles.

- Facilitation de la communication : Le NLP vise à construire des systèmes qui aident les humains à communiquer plus efficacement. Cela peut être entre eux ou avec des dispositifs électroniques. Par exemple, grâce au NLP, nous pouvons avoir des assistants virtuels qui comprennent nos demandes vocales et y répondent.

- Systèmes utiles : Le NLP permet de développer des systèmes utiles tels que la classification de textes, la synthèse automatique de textes, la reconnaissance vocale, et bien d'autres applications qui facilitent notre quotidien.

- - Briser les barrières linguistiques : Avec le NLP, il est possible de communiquer sans barrière linguistique. Les systèmes de traduction automatique, par exemple, permettent de traduire instantanément du texte ou de la parole d'une langue à une autre.

- Modélisation et analyse des propriétés du langage : Le NLP permet également de modéliser et d'analyser les propriétés intrinsèques du langage, ce qui est essentiel pour comprendre comment les humains utilisent le langage pour communiquer.

### b. Quelles sont les principales étapes du traitement du langage naturel ? Expliquez brièvement chacune d'entre elles.

Les principales étapes du traitement du langage naturel (NLP) sont souvent regroupées dans ce que l'on appelle le "NLP Pipeline".

- Tokenisation : C'est le processus de découpage du texte en unités plus petites, appelées tokens. Chaque mot, phrase ou autre unité significative est considéré comme un token.

- Lemmatisation : Cette étape permet de garder la racine du mot, regroupant ainsi plusieurs variantes d'un même mot en une seule forme canonique. Par exemple, "courant" et "couru" peuvent être réduits à la racine "cour".

- Stopwords : Il s'agit de mots couramment utilisés qui sont généralement ignorés lors de l'analyse car ils n'apportent pas de valeur significative (ex : "et", "le", "à").

- Morphologie : Cette étape se concentre sur la structure des mots, identifiant des éléments tels que les racines, les préfixes et les suffixes.

- Bag of Words (Sac de mots) : C'est une représentation du texte où chaque mot est représenté par un indicateur (souvent une fréquence) dans un vecteur. Les informations sur l'ordre des mots sont généralement perdues.

- Étiquetage grammatical (POS tagging) : Chaque mot est étiqueté avec sa partie du discours (nom, verbe, adjectif, etc.).

- Analyse syntaxique : Cette étape vise à déterminer la structure grammaticale d'une phrase, en identifiant les sujets, les objets, etc.

- Analyse sémantique : Ici, on détermine le sens des mots dans le contexte et la signification globale de la phrase.

- Analyse du discours : Cette étape se concentre sur la compréhension de l'ensemble du discours ou du texte, en identifiant comment les phrases s'articulent entre elles.

- Reconnaissance d'entités nommées : Identification et classification des entités (noms propres, lieux, dates, etc.) dans le texte.

- Désambiguïsation lexicale : Sélection du sens approprié d'un mot basé sur le contexte.

- Analyse des sentiments : Détermination de l'émotion ou de l'opinion exprimée dans le texte.

- Traduction automatique, génération de texte, etc. : Ces étapes avancées utilisent toutes les informations précédemment extraites pour accomplir des tâches spécifiques.

### c. Quelles sont les principales applications du NLP dans le monde réel ? Donnez des exemples concrets.

Les principales applications du traitement du langage naturel (NLP) dans le monde réel comprennent :

- Systèmes de réponse aux questions : Ces systèmes sont conçus pour répondre aux questions posées en langage naturel. Ils sont souvent utilisés dans les assistants virtuels et les chatbots. Par exemple, des plateformes comme Siri, Alexa ou Google Assistant utilisent le NLP pour comprendre et répondre aux requêtes des utilisateurs.

- Classification de texte : Le NLP permet de catégoriser automatiquement des textes dans différentes classes ou catégories. Cela est utile, par exemple, pour trier des emails, classer des articles de presse par sujet, ou encore pour la modération de contenu.

- Résumé automatique de texte : Cette application vise à produire un résumé concis d'un texte plus long, en conservant les informations principales.

- Traduction automatique : Le NLP est à la base des systèmes de traduction comme Google Translate, qui traduisent du texte d'une langue à une autre.

- Reconnaissance vocale : Le NLP permet de convertir la parole en texte, et est utilisé dans des applications comme la dictée vocale ou la commande vocale.

- Assistants personnels numériques : Ces systèmes, comme Siri ou Alexa, utilisent le NLP pour comprendre les requêtes vocales des utilisateurs et y répondre.

- Analyse des sentiments : Le NLP est utilisé pour déterminer si un texte exprime un sentiment positif, négatif ou neutre. C'est particulièrement utile pour les entreprises qui souhaitent surveiller la perception de leur marque sur les réseaux sociaux.

- Correction orthographique et grammaticale : Des outils comme Grammarly utilisent le NLP pour détecter et corriger les erreurs dans le texte.

- Recherche d'information : Le NLP améliore la pertinence des résultats de recherche en comprenant le contexte et la signification des requêtes.

- Chatbots et agents de service client : Ces systèmes interagissent avec les utilisateurs en langage naturel pour répondre à leurs questions ou résoudre leurs problèmes.



### d. Quels sont les défis courants auxquels les modèles de NLP sont confrontés ? Comment le biais peut-il affecter ces modèles ?

Les défis courants auxquels les modèles de NLP sont confrontés comprennent :

- Ambiguïté : Les langues sont naturellement ambiguës. Un même mot ou une même phrase peut avoir plusieurs significations en fonction du contexte. Par exemple, le mot "banque" peut désigner une institution financière ou le bord d'une rivière.

- Variabilité : Il existe de nombreuses façons d'exprimer la même idée ou le même sentiment en langage naturel. Les gens utilisent différents styles, dialectes, argots et structures grammaticales pour communiquer la même information.

- Langage non statique : Le langage évolue constamment. De nouveaux mots, expressions et argots sont régulièrement ajoutés à la langue. Par exemple, des mots comme "google" sont devenus des verbes, et des expressions comme "#TBT" (throwback Thursday) ou "DM" (direct message) ont acquis des significations spécifiques avec l'essor des médias sociaux.

Les modèles de NLP peuvent hériter des biais présents dans les données sur lesquelles ils sont formés, ce qui peut conduire à des résultats inéquitables ou incorrects.

### e. Comment évaluez-vous la performance d'un modèle de NLP ? Quelles sont les métriques couramment utilisées ?

L'évaluation de la performance d'un modèle de NLP dépend de la tâche spécifique pour laquelle le modèle est conçu. Voici quelques méthodes couramment utilisées pour évaluer différents types de modèles de NLP :

- Précision, Rappel et Score F1 : Ces métriques sont couramment utilisées pour évaluer les modèles de classification. La précision mesure la proportion de prédictions positives qui sont réellement positives, le rappel mesure la proportion de vrais positifs qui sont correctement identifiés, et le score F1 est la moyenne harmonique de la précision et du rappel.

- Erreur quadratique moyenne (MSE) : Pour les tâches de régression, comme la prédiction du score de sentiment, le MSE mesure la différence entre les valeurs prédites et les valeurs réelles.

- BLEU, ROUGE, METEOR : Ces métriques sont utilisées pour évaluer les modèles de traduction automatique ou de génération de texte. Elles comparent la sortie du modèle à une ou plusieurs références pour déterminer la qualité de la traduction ou du texte généré.

- Perplexité : Utilisée pour évaluer les modèles de langage, la perplexité mesure la capacité du modèle à prédire l'échantillon de test. Une faible perplexité indique que le modèle est bon pour prédire l'échantillon.

- Taux d'erreur de mot (WER) : Pour les modèles de reconnaissance vocale, le WER mesure le nombre de mots mal reconnus par rapport au nombre total de mots.

- Évaluation qualitative : Parfois, une évaluation qualitative est nécessaire, où des humains évaluent la qualité ou la pertinence des sorties du modèle. Cela peut être particulièrement utile pour des tâches comme la génération de texte ou la réponse aux questions.

- Courbes ROC et AUC : Ces métriques évaluent la performance des modèles de classification en fonction de différents seuils de décision.

## II. Modélisation

### 1. télécharger les données sur les avis des clients dans un restaurant

In [ ]:
import numpy as np
import pandas as pd

In [ ]:
df = pd.read_csv("/content/Restaurant_Reviews.tsv",delimiter='\t');
df.head()

Review  Liked
0                           Wow... Loved this place.      1
1                                 Crust is not good.      0
2          Not tasty and the texture was just nasty.      0
3  Stopped by during the late May bank holiday of...      1
4  The selection on the menu was great and so wer...      1

### 2. Faire le prétraitement de ces données et expliquer l’importance de chaque étape

In [ ]:
print(df.columns)
print(df.shape)
print(df.isna().sum())

Index(['Review', 'Liked'], dtype='object')
(1000, 2)
Review    0
Liked     0
dtype: int64


In [37]:
import spacy
import nltk
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer
from nltk.corpus.reader import wordnet
from nltk import pos_tag
from string import punctuation

In [38]:
# Chargement du modèle spacy
nlp = spacy.load('en_core_web_sm')

In [39]:
# Téléchargement des ressources nécessaires
nltk.download('stopwords')
nltk.download('wordnet')
stop_words = set(stopwords.words('english'))

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


In [40]:
from nltk.stem.snowball import SnowballStemmer
from nltk.stem import WordNetLemmatizer

In [45]:
# Fonction de prétraitement
def preprocess(text):
    # Tokenisation
    tokens = [i.text for i in nlp(text)]

    # Suppression de la ponctuation
    tokens_no_punct = [token for token in tokens if token not in punctuation]

    # Suppression des stopwords
    tokens_no_stopwords = [token for token in tokens_no_punct if token not in stop_words]

    # Lemmatisation avec POS tagging
    lemmatizer = WordNetLemmatizer()
    lemmatized_tokens = [lemmatizer.lemmatize(i) for i in tokens_no_stopwords]

    # Stemming
    stemmer = SnowballStemmer("english")
    stemmed_tokens = [stemmer.stem(token) for token in lemmatized_tokens]

    return ' '.join(stemmed_tokens)

In [46]:
df['Processed_Review'] = df['Review'].apply(preprocess)

In [47]:
df['Processed_Review'].head()

0                                       wow love place
1                                           crust good
2                                   tasti textur nasti
3    stop late may bank holiday rick steve recommen...
4                              select menu great price
Name: Processed_Review, dtype: object

Explications :
- Tokenisation :
C'est le processus de découpage du texte en unités plus petites, appelées tokens. Chaque mot, phrase ou autre unité significative est considéré comme un token. Cette étape est essentielle car elle transforme le texte brut en une liste de mots qui peuvent être analysés et traités.

- Lemmatisation :
Elle permet de garder la racine du mot, regroupant ainsi plusieurs variantes d'un même mot en une seule forme canonique. Par exemple, "courant" et "couru" peuvent être réduits à la racine "cour". La lemmatisation est préférable au stemming car elle tient compte du contexte et convertit le mot en sa forme de base significative.

- Stopwords :
Il s'agit de mots couramment utilisés qui sont généralement ignorés lors de l'analyse car ils n'apportent pas de valeur significative (ex : "et", "le", "à"). Leur suppression réduit la dimensionnalité des données et accélère le processus d'entraînement.

- Morphologie :
Cette étape se concentre sur la structure des mots, identifiant des éléments tels que les racines, les préfixes et les suffixes. Elle aide à comprendre la formation des mots et peut être utilisée pour extraire des caractéristiques pertinentes du texte.

- Bag of Words (Sac de mots) :
C'est une représentation du texte où chaque mot est représenté par un indicateur (souvent une fréquence) dans un vecteur. Les informations sur l'ordre des mots sont généralement perdues. Cette technique transforme le texte en une forme numérique qui peut être utilisée pour entraîner des modèles.

In [48]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.feature_extraction.text import CountVectorizer

In [49]:
# Appliquer le prétraitement à toutes les critiques
df['Processed_Review'] = df['Review'].apply(preprocess)

# Conversion des critiques prétraitées en représentation Bag of Words
cv = CountVectorizer()
X = cv.fit_transform(df['Processed_Review']).toarray()
y = df['Liked'].values

# Division des données en ensembles d'entraînement et de test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)


### 3. Proposer un modèle en expliquant sa structure, son architecture et son fonctionnement

In [50]:
# Initialisation et entraînement du modèle
classifier = LogisticRegression(random_state=0)

- Structure et Architecture :

La régression logistique est un modèle statistique utilisé pour prédire la probabilité d'une variable dépendante catégorielle. Dans le contexte de la classification binaire, la variable dépendante peut prendre deux valeurs : 0 ou 1.

L'architecture de la régression logistique est simple. Elle prend en entrée des caractéristiques (dans notre cas, les mots ou les tokens du texte) et les pèse à l'aide de coefficients pour produire une somme pondérée. Cette somme est ensuite passée à travers une fonction logistique (ou sigmoïde) pour produire une sortie entre 0 et 1.

- Fonctionnement :

Somme pondérée : La première étape consiste à calculer une somme pondérée des caractéristiques.

Fonction Sigmoïde : La somme pondérée est ensuite passée à travers la fonction sigmoïde pour la contraindre entre 0 et 1.
Le résultat est une probabilité qui reflète la probabilité que la sortie soit de classe 1.

Seuil : Enfin, un seuil (généralement 0,5) est appliqué à la probabilité pour produire une prédiction finale de 0 ou 1.

Entraînement : Le modèle est entraîné en utilisant une fonction de coût, généralement la log-perte, pour ajuster les poids et le biais de manière à minimiser l'erreur entre les prédictions et les vraies valeurs.



### 4. Faire l'entrainement du modèle

In [51]:
classifier.fit(X_train, y_train)

LogisticRegression(random_state=0)

### 5. Proposer une mesure évaluation de votre modèle et expliquer les résultats

In [52]:
from sklearn.metrics import accuracy_score, classification_report

# Prédiction sur l'ensemble de test
y_pred = classifier.predict(X_test)

# Calcul de la précision
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy*100:.2f}%")

# Rapport de classification
report = classification_report(y_test, y_pred)
print(report)


Accuracy: 72.00%
              precision    recall  f1-score   support

           0       0.69      0.77      0.73        97
           1       0.76      0.67      0.71       103

    accuracy                           0.72       200
   macro avg       0.72      0.72      0.72       200
weighted avg       0.72      0.72      0.72       200



- Accuracy (Précision) :
La précision est de 72,00%. Cela signifie que 72% des prédictions faites par le modèle sur l'ensemble de test étaient correctes. C'est une mesure générale de la performance du modèle.

- Precision (Précision) :
Pour la classe 0 : 0,69 ou 69%. Cela signifie que, parmi toutes les prédictions où le modèle a prédit la classe 0, 69% étaient effectivement de la classe 0.
Pour la classe 1 : 0,76 ou 76%. Cela signifie que, parmi toutes les prédictions où le modèle a prédit la classe 1, 76% étaient effectivement de la classe 1.

- Recall (Rappel) :
Pour la classe 0 : 0,77 ou 77%. Cela signifie que, parmi tous les vrais exemples de la classe 0 dans l'ensemble de test, le modèle a correctement identifié 77% d'entre eux.
Pour la classe 1 : 0,67 ou 67%. Cela signifie que, parmi tous les vrais exemples de la classe 1 dans l'ensemble de test, le modèle a correctement identifié 67% d'entre eux.

- F1-Score :
Le F1-Score est la moyenne harmonique de la précision et du rappel. Il donne une mesure plus complète des performances du modèle, en tenant compte à la fois de la précision et du rappel.
Pour la classe 0 : 0,73 ou 73%.
Pour la classe 1 : 0,71 ou 71%.

En résumé, votre modèle a une performance générale de 72%. Il semble être légèrement meilleur pour prédire la classe 1 par rapport à la classe 0 en termes de précision, mais son rappel pour la classe 1 est inférieur à celui de la classe 0. Cela pourrait indiquer que le modèle manque certains exemples positifs (classe 1) tout en étant plus prudent lors de la prédiction de la classe positive.